api/v2alpha1/helmrelease_types.go

```diff
@@ -563,10 +563,18 @@ func (in Uninstall) GetTimeout(defaultTimeout metav1.Duration) metav1.Duration {
 
 // HelmReleaseStatus defines the observed state of HelmRelease
 type HelmReleaseStatus struct {
-	// ObservedGeneration is the last reconciled generation.
+	// ObservedGeneration is the last observed generation.
 	// +optional
 	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
 
+	// ObservedStateReconciled represents whether the observed state has been successfully reconciled.
+	// +optional
+	ObservedStateReconciled bool `json:"observedStateReconciled,omitempty"`
```

Comment:
Isn't this redundant since a diff of `ObservedGeneration` would give you the same info?

Reply:
A failure will still bump the `ObservedGeneration` while the `ObservedStateReconciled` only moves after a successful reconciliation.

Reply:
So `ObservedStateReconciled` can be composed from ObservedGeneration and Ready=true?

Reply:
The problem with using Ready=true is that it can later again become false due to e.g. transient k8s API failures, source-controller failures, failures in dependsOn resolution etc, but then after those resolve, as long as no new state is observed for the HelmRelease we still want to remember that we have already reconciled the existing state, to avoid spurious upgrades.

Reply:
To clarify the "observed state" includes not just the internal ObservedGeneration state, but also external state, namely the resolved chart revision, the values checksum (from ConfigMaps / Secrets), and the last release revision number so that out of band release management re-triggers reconciliation.

Reply:
I think the main alternative to "remembering" that we've already applied the state is to actually look up the state that has been applied and diff against it, as was done in the Helm Operator via a dry run, but I think there were too many problems with this and so it was decided against for the helm-controller. I personally like the simplicity gained with this approach. I think the only thing you really lose is the ability to detect drift in the rendered resources, which is not a use case I've personal had.

Reply:
Ok let's revisit this when we'll design the drift detection.